### PR TITLE
🐛 Fix biospecimens and diagnoses links by diagnosis_id and biospecimen_id

### PR DESCRIPTION
**Pull request by parimalak**
issue : /diagnoses?biospecimen_id=BS_id ,  /biospecimen?diagnosis_id=DG_idresulted in internal server error
fix :  Added queries to join the biospecimens/diagnoses with BiospecimenDiagnosis

### CODE REVIEW

File: tests/test_pagination.py [hunk 1]
```diff
@@ -397,3 +396,27 @@ def _check_link(self, link_str, params):
         assert float(q_params.get('after')[0])
         for k, v in params.items():
             assert q_params.get(k)[0] == v
+
+    @pytest.mark.parametrize('endpoint,expected_total', [
+        ('/biospecimens', 1),
+        ('/diagnoses', 1),
+    ])
+    def test_all_filter(self, client, participants,
```
Comment (dankolbman) [inline tests/test_pagination.py]: This test would be better fit for the test_resources file for the two types.

File: dataservice/api/diagnosis/resources.py [hunk 2]
```diff
@@ -38,16 +38,26 @@ def get(self, filter_params, after, limit):
...
-        if study_id:
+        if biospecimen_id:
+            q = q.join(Diagnosis.biospecimens).\
+                filter(BiospecimenDiagnosis.biospecimen_id == biospecimen_id)
+
+        # Apply study_id filter and biospecimen_id filter
+        if biospecimen_id and study_id:
+            q = q.join(Diagnosis.biospecimens).\
+                filter(BiospecimenDiagnosis.biospecimen_id == biospecimen_id).\
+                filter(Participant.study_id == study_id)
```
Comment (dankolbman) [inline dataservice/api/diagnosis/resources.py]: Still not sure I understand how this works.

File: dataservice/api/diagnosis/resources.py [hunk 3]
```diff
@@ -38,16 +38,26 @@ def get(self, filter_params, after, limit):
...
-        # Apply study_id filter
-        from dataservice.api.participant.models import Participant
-        if study_id:
+        if biospecimen_id:
+            q = q.join(Diagnosis.biospecimens).\
+                filter(BiospecimenDiagnosis.biospecimen_id == biospecimen_id)
+
+        # Apply study_id filter and biospecimen_id filter
+        if biospecimen_id and study_id:
+            q = q.join(Diagnosis.biospecimens).\
```
Comment (dankolbman) [inline dataservice/api/diagnosis/resources.py]: I get this warning. You may not have to join here:
```
SAWarning: Pathed join target Diagnosis.biospecimens has already been joined to; skipping
```

File: dataservice/api/diagnosis/resources.py [hunk 4]
```diff
@@ -38,16 +38,23 @@ def get(self, filter_params, after, limit):
...
+        if biospecimen_id and study_id:
+            q = q.join(Participant.diagnoses).\
+                join(Diagnosis.biospecimens).\
+                filter(BiospecimenDiagnosis.biospecimen_id == biospecimen_id).\
+                filter(Participant.study_id == study_id)
+        elif study_id and not biospecimen_id:
             q = (q.join(Participant.diagnoses)
                  .filter(Participant.study_id == study_id))
-
+        elif biospecimen_id and not study_id:
```
Comment (dankolbman) [inline dataservice/api/diagnosis/resources.py]: I think the first if could be removed and then just do
```
if study_id:
    ...
if biospecimen_id:
    ...
```

File: dataservice/api/biospecimen/resources.py [hunk 5]
```diff
@@ -37,18 +38,27 @@ def get(self, filter_params, after, limit):
...
+        diagnosis_id = filter_params.pop('diagnosis_id', None)
         # Apply filter params
         q = (Biospecimen.query
              .filter_by(**filter_params))
 
-        # Apply study filter param
         from dataservice.api.participant.models import Participant
-        if study_id:
+        # Apply study_id filter and diagnosis_id filter
+        if diagnosis_id and study_id:
```
Comment (dankolbman) [inline dataservice/api/biospecimen/resources.py]: You should be able to remove this and change the two `elif`s below to `if`s

Comment (parimalak) [reply dataservice/api/biospecimen/resources.py]: all the three conditions have different joins.  if we remove the `elifs`  two of the conditions will be true at the same time and the query throws multiple joins on the same table.

Comment (dankolbman) [reply dataservice/api/biospecimen/resources.py]: I think they are doing the same joins though?

Comment (parimalak) [reply dataservice/api/biospecimen/resources.py]: ahh, I see what you are saying if both conditions are satisfied both the conditions queries will be executed.

**Review by dankolbman — APPROVED** [review]
looks good!

**Review by znatty22 — APPROVED** [review]
👍

**Pull request merged**